### PR TITLE
Add update support for bigtable clusters (such as resizing)

### PR DESCRIPTION
**Pull request by stankiewicz**
This PR is to address [Issue #2521](https://github.com/terraform-providers/terraform-provider-google/issues/2521) for terraform-google-provider 

What will be doable is scaling up and down number of clusters within instance without recreate. 
If anything else is changed within cluster (change of disk, zone, name) it will be forced to recreate, so the behaviour mimics what is possible in UI.

fyi: 
- this is my *first* PR in this project
- this is my *first* golang chunk of code
- I'm new to TF

I've tested it with all kind of scenarios (resize, force recreate, both)

Any feedback highly appreciated so I can complete this PR.

### CODE REVIEW

**Comment by stankiewicz** [comment]
Will add test for this case.

**Comment by stankiewicz** [comment]
One challenge I have,
Error: Provider produced inconsistent final plan

       " When expanding the plan for google_bigtable_instance.instance to include new
        values learned so far during apply, provider "google" produced an invalid new
        value for .cluster[1].zone: was cty.StringVal("us-central1-b"), but now
        cty.StringVal("us-central1-c")."

File: third_party/terraform/resources/resource_bigtable_instance.go [hunk 1]
```diff
@@ -335,29 +374,43 @@ func resourceBigtableInstanceClusterReorderTypeList(diff *schema.ResourceDiff, m
...
+			c := clusters[keys[0].Interface().(string)]
+			delete(clusters, keys[0].Interface().(string))
 			old_cluster_order = append(old_cluster_order, c)
 		}
 	}
+	for _, value := range clusters {
+		old_cluster_order = append(old_cluster_order, value)
+	}
 
 	err := diff.SetNew("cluster", old_cluster_order)
```
Comment (stankiewicz) [inline third_party/terraform/resources/resource_bigtable_instance.go]: this is producing "Provider produced inconsistent final plan" after the change.

**Comment by rileykarson** [comment]
This feature is being tracked at a new issue, https://github.com/terraform-providers/terraform-provider-google/issues/4318. (Sorry for the web of BT cluster updating issues! I think I missed an xref between them somewhere.) I've been working with a member of the BT team to have this functionality added directly to the BT client library as https://code-review.googlesource.com/c/gocloud/+/44410, which should land soon.

I'd strongly prefer we use that implementation instead of a separate one in TF. It's a commitment of mine to add it to the provider once that change lands. Given that, I'll probably take over the implementation unless you'd prefer to make the contribution.

FWIW, the error you encountered is because Terraform `0.12` and onwards are _very_ strict that values shown in `terraform plan` match the ordering of values before/after `terraform apply`. I suspect that ordering in the `CustomizeDiff` is different than the real ordering applied in `d.Set`, causing the issue you encountered.

**Comment by stankiewicz** [comment]
@rileykarson  I've spoken to Gary about that. If his change can land soon then great! 
I think that in that case you can take over integration or I can help under your guidance if you don't have capacity. I have some and some work is done here already as same logic will need to be implemented. I know where is the issue but I don't know yet how to fix it.

**Comment by stankiewicz** [comment]
@rileykarson , I've updated code and tests, so I don't have any more "inconsistent final plan".

**Comment by rileykarson** [comment]
Sorry for the wait- I think https://code-review.googlesource.com/c/gocloud/+/44410 is close to merging.

File: third_party/terraform/resources/resource_bigtable_instance.go [hunk 2]
```diff
@@ -366,3 +409,53 @@ func resourceBigtableInstanceClusterReorderTypeList(diff *schema.ResourceDiff, m
...
+func resourceResize(diff *schema.ResourceDiff, meta interface{}) error {
+	old_count, new_count := diff.GetChange("cluster.#")
+	if old_count.(int) != new_count.(int) {
+		log.Printf("[DEBUG] resize from %d to %d", old_count, new_count)
+	}
+
+	new_clusters := make(map[string]interface{}, new_count.(int))
+	old_clusters := make(map[string]interface{}, old_count.(int))
+
+	for i := 0; i < new_count.(int) || i < old_count.(int); i++ {
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_bigtable_instance.go]: Do you mind commenting the blocks like this with what they're doing? While it's possible to figure out, customizeDiffs are pretty finicky and it's a big help to be able to read quickly what they're doing.

File: third_party/terraform/tests/resource_bigtable_instance_test.go [hunk 3]
```diff
@@ -75,6 +75,41 @@ func TestAccBigtableInstance_cluster(t *testing.T) {
...
+	instanceName := fmt.Sprintf("tf-test-%s", acctest.RandString(10))
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckBigtableInstanceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccBigtableInstance_one_in_the_cluster(instanceName, 3),
+				Check: resource.ComposeTestCheckFunc(
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_bigtable_instance_test.go]: After rebasing, you'll probably need to change to use "import-style" checks. See https://github.com/GoogleCloudPlatform/magic-modules/pull/2409/files#diff-1296b3b7cf174ed6b01d0139e46bdd24L69-R76.

If you're seeing differences on `cluster` in "ImportStateVerify" that you don't expect, flag them here and don't worry about fixing them before pushing.

**Comment by rileykarson** [comment]
Superseded with https://github.com/GoogleCloudPlatform/magic-modules/pull/2923

Thanks for taking the time to submit this! Sorry about the unfortunate timing between your work + the client lib work I'd been waiting to integrate, you did a great job with your changes.

**Pull request closed without merging**